awscli/customizations/history/db.py

```diff
@@ -0,0 +1,244 @@
+        self._connection.row_factory = row_factory
+
+
+class PayloadSerializer(json.JSONEncoder):
+    def _encode_mutable_mapping(self, obj):
+        return dict(obj)
+
+    def _encode_datetime(self, obj):
+        return obj.isoformat()
+
+    def _unknown(self, obj, type_name):
+        return type_name
+
+    def default(self, obj):
+        if isinstance(obj, datetime.datetime):
+            return self._encode_datetime(obj)
+        elif isinstance(obj, MutableMapping):
+            return self._encode_mutable_mapping(obj)
+        else:
+            return repr(obj)
+
+
+class DatabaseRecordWriter(object):
+    _WRITE_RECORD = """
+    INSERT INTO records(id, request_id, source, event_type, timestamp, payload)
+    VALUES (?,?,?,?,?,?)"""
+
+    def __init__(self, connection=None):
+        if connection is None:
+            connection = DatabaseConnection()
+        self._connection = connection
+
+    def write_record(self, record):
+        # This method is not threadsafe by itself, it is only threadsafe when
+        # used inside a handler bound to the HistoryRecorder in botocore which
+        # is protected by a lock.
+        db_record = self._create_db_record(record)
+        self._connection.execute(self._WRITE_RECORD, db_record)
+
+    def _create_db_record(self, record):
+        event_type = record['event_type']
+        json_serialized_payload = json.dumps(record['payload'],
+                                             cls=PayloadSerializer)
+        db_record = (
+            record['command_id'],
+            record.get('request_id'),
+            record['source'],
+            event_type,
+            record['timestamp'],
+            json_serialized_payload
+        )
+        return db_record
+
+
+class DatabaseRecordReader(object):
+    _ORDERING = 'ORDER BY timestamp'
+    _GET_LAST_ID_RECORDS = """
+    SELECT * FROM records
+    WHERE id =
+    (SELECT id FROM records WHERE timestamp =
+    (SELECT max(timestamp) FROM records)) %s;""" % _ORDERING
+    _GET_RECORDS_BY_ID = 'SELECT * from records where id = ? %s' % _ORDERING
+
+    def __init__(self, connection=None):
+        if connection is None:
+            connection = DatabaseConnection()
+        self._connection = connection
+        self._connection.row_factory = self._row_factory
+
+    def _row_factory(self, cursor, row):
+        d = {}
+        for idx, col in enumerate(cursor.description):
+            val = row[idx]
+            if col[0] == 'payload':
+                val = json.loads(val)
+            d[col[0]] = val
+        return d
+
+    def iter_latest_records(self):
+        cursor = self._connection.cursor()
+        cursor.execute(self._GET_LAST_ID_RECORDS)
+        for row in cursor:
+            yield row
+
+    def iter_records(self, record_id):
+        cursor = self._connection.cursor()
+        cursor.execute(self._GET_RECORDS_BY_ID, [record_id])
+        for row in cursor:
+            yield row
+
+
+class RecordBuilder(object):
```

Comment:
Did not see really a test class for the `RecordBuilder`. Do you think it makes sense to add some specifically related to making sure everything is formatted correctly, it handles payloads properly, and injects timestamps? It looks like you really have a lot of logic related to request_id's but not so much the other aspects of it.